docs/interpreter_checklist.md

```diff
@@ -0,0 +1,60 @@
+
+In this document, we summarize the guidelines while implementing and reviewing
+interpreter for an op. We have intentionally tied a few auxiliary action items,
+related to verifier and type inference, with the idea of making progress in
+those fronts alongside the interpreter implementation.
+
+1. While implementing the interpreter for an op, double check with related
+   implementations (e.g.
+   <!-- markdownlint-disable line-length -->
+   [hlo_interpreter](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/hlo/evaluator)
+   <!-- markdownlint-enable line-length -->
+   and
+   <!-- markdownlint-disable line-length -->
+   [mlir_interpreter](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/xla/mlir_hlo/tools/mlir_interpreter)).
+   <!-- markdownlint-enable line-length -->
+   This action item will allow mutual improvements.
+1. Replace the comments, related to input/output constraints, in
+   the implementations, specifically for verifiers and type inference, with
+   "constraint-labels" (e.g. `Cn` in
+   <!-- markdownlint-disable line-length -->
+   [here](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#constraints-2) )
+   <!-- markdownlint-enable line-length -->
+   used in the
+   <!-- markdownlint-disable line-length -->
+   [specification](https://github.com/openxla/stablehlo/blob/main/docs/spec.md).
+   <!-- markdownlint-enable line-length -->
+   This is to explicitly align all the implementations with the specification:
+   the single source of truth, and help to track gaps in those implementations
+   back to the spec.
+   1. It is OK to have a comment with multiple constraint-labels or to have
+      multiple comments with the same constraint-label. It all depends on how
+      the constraint-label(s) are coded in the implementations.
+   1. In case there is a mismatch between the constraints in the
+      implementation VS those in the specification, make sure there is an
+      open issue reflecting that discrepancy.
+1. Make sure that the there are sufficient tests to cover for the
+   implementations (e.g. verifiers, type inference, interpreter) of the
+   specification. Specifically,
+   1. For interpreter: Write tests following the
+      <!-- markdownlint-disable line-length -->
+      [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/reference.md#testing-guidelines).
+      <!-- markdownlint-enable line-length -->
+   1. For verifier and type inference: Make sure that there is at least one test
+      for each constraint in verifier and type inference methods. These tests
+      will mostly be negative tests exercising the fact that the constraints are
+      not met. Make sure to add at least one positive tests in case there are
+      none. Note that the action item here is not related to tests for checking
+      if the types are correctly inferred or not, which we believe is already
+      done with sufficient thoroughness.
```

Comment:
* Group tests by ops alphabetically and before every op add a comment section with the ops name?
  * Within a single op, order the tests in some manner like positive test followed by negative tests?
  * Maybe write the constraints `Cn` in comments in tests for testing constraints?

This will make the tests also organised, making sure all the cases are extensibly and exhaustively covered and tie it to the spec.

Reply:
@sdasgup3

Reply:
#123

Reply:
Nice suggestion!
I agree that there is a lot opportunity to to sanitize https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/ops_stablehlo.mlir and I have opened a [ticket](https://github.com/openxla/stablehlo/issues/993) enumerating some of my thoughts on (1) and related items. 

The action items, related to testing verifier and type inferencers, are a few auxiliary things that we want to accomplish at a per op level.  Given that the scope of (1) is beyond "per op" level, we can address this using the above ticket. For other I have updated the guidelines. PTAL